titan-compiler/checker.lua

```diff
@@ -689,43 +621,103 @@ local function checkfunc(node, st, errors)
     end
 end
 
--- Third typechecking pass over the module, checks function bodies
+-- Checks function bodies
 --   ast: AST for the whole module
 --   st: symbol table
 --   errors: list of compile-time errors
-local function thirdpass(ast, st, errors)
-    for _, tlnode in ipairs(ast) do
-        if not tlnode._ignore then
-            local tag = tlnode._tag
-            if tag == "TopLevel_Func" then
-                st:with_block(checkfunc, tlnode, st, errors)
+local function checkbodies(ast, st, errors)
+    for _, node in ipairs(ast) do
+        if not node._ignore then
+            if node._tag == "TopLevel_Func" then
+                st:with_block(checkfunc, node, st, errors)
             end
         end
     end
 end
 
--- Gets type information for all imported modules and puts
--- it in the symbol table
-local function importpass(ast, st, errors, loader)
-    for _, tlnode in ipairs(ast) do
-        if tlnode._tag == "TopLevel_Import" then
-            local name = tlnode.modname
-            if st:find_dup(tlnode.localname) then
-                typeerror(errors, "duplicate declaration for " .. tlnode.localname, tlnode._pos)
-                tlnode._ignore = true
-            else
-                local modtype, errs = checker.checkimport(name, loader)
-                if modtype then
-                    tlnode._type = modtype
-                    for _, err in ipairs(errs) do
-                        table.insert(errors, err)
-                    end
-                    st:add_symbol(tlnode.localname, tlnode)
-                else
-                    tlnode._type = types.Nil
-                    typeerror(errors, "problem loading module '%s': %s", tlnode._pos, name, errs)
+-- Return the name given the toplevel node
+local function toplevel_name(node)
+    local tag = node._tag
+    if tag == "TopLevel_Import" then
+        return node.localname
+    elseif tag == "TopLevel_Var" then
+        return node.decl.name
+    elseif tag == "TopLevel_Func" then
+        return node.name
+    else
+        error("tag not found " .. tag)
+    end
+end
+
+-- Typecheck the toplevel node
+local tlcheckers = {
```

Comment:
I think the whole typechecker should follow this pattern. It is easier to merge common parts in the code.

Reply:
I'm a fan of more table-based (in a Titan-future object-based?) approach like this rather than long sequences of if-else-if. And it's something we can transition to gradually (even if it results in less "consistency" in the meantime). I'm a fan of incremental improvement!